File: mmd_generator/transformation_log.py
"""Transformation audit trail (Section 36)."""
from __future__ import annotations

from typing import List

from .models import TransformationEntry


class TransformationLog:
    """Records every raw -> MMD transformation the generator makes, in
    order, so the process is fully auditable after the fact."""

    def __init__(self) -> None:
        self._entries: List[TransformationEntry] = []

    def record(self, rule: str, raw: str, became: str, reason: str = "") -> None:
        self._entries.append(
            TransformationEntry(index=len(self._entries) + 1, rule=rule, raw=raw, became=became, reason=reason)
        )

    def entries(self) -> List[TransformationEntry]:
        return list(self._entries)

    def to_text(self) -> str:
        lines = []
        for e in self._entries:
            line = f"[{e.index}] ({e.rule}) RAW: {e.raw!r} -> BECAME: {e.became!r}"
            if e.reason:
                line += f"  # {e.reason}"
            lines.append(line)
        return "\n".join(lines)

File: mmd-dashboard/lib/transformationLog.ts
// Transformation audit trail (Section 36). Port of transformation_log.py.

import type { TransformationEntry } from "./models";

export class TransformationLog {
  private entriesList: TransformationEntry[] = [];

  record(rule: string, raw: string, became: string, reason = ""): void {
    this.entriesList.push({
      index: this.entriesList.length + 1,
      rule,
      raw,
      became,
      reason: reason || undefined,
    });
  }

  entries(): TransformationEntry[] {
    return [...this.entriesList];
  }

  toText(): string {
    return this.entriesList
      .map((e) => {
        let line = `[${e.index}] (${e.rule}) RAW: ${JSON.stringify(e.raw)} -> BECAME: ${JSON.stringify(e.became)}`;
        if (e.reason) line += `  # ${e.reason}`;
        return line;
      })
      .join("\n");
  }
}
